### ISSUE TITLE
Bump go mod version from 1.22 to 1.22.0 to fix toolchain download issue

### ISSUE THREAD

**Issue by cmoulliard** [buildpacks/pack#2267]
### Description

TL&DR
Pack cannot be build in an hermetic and containerized environment as the version of go `1.22` declared within the go.mod file do not correspond to a toolchain version `1.22.0`

----

Such a limitation comes from Golang itself how it processes go.mod files. When Go sees the language version, but not a toolchain version, i.e. `1.22` vs `1.22.0` it's not going to be able to download a correct toolchain for the given language version. Go community is aware of that problem and they fixed it in `1.23` so that the inference of the `.0`th toolchain from a particular language version works. 

### Proposed solution
Bump the go version defined within the go.mod file

### Additional context

See error reported when building pack in an hermetic and containerized environment
![Screenshot 2024-09-26 at 19 31 43](https://github.com/user-attachments/assets/6ba3c279-ba00-458b-8a1a-6f318316da7d)

**Comment by jjbustamante** [comment]
@cmoulliard I have a question about this.

If we bump the go mod version to `1.22.0`  will this affect the workflow version when we are compiling the tool [here](https://github.com/buildpacks/pack/blob/main/.github/workflows/build.yml#L189-L192)? If I am not wrong, we are using the `check-latest` option which uses the latest go fix version `1.22.x`. 

For example, on my machine, I can see the latest pack version was compiled using go `1.22.5`

```bash
>$ which pack
/home/linuxbrew/.linuxbrew/bin/pack
>$ go version /home/linuxbrew/.linuxbrew/bin/pack
/home/linuxbrew/.linuxbrew/bin/pack: go1.22.5
>$ pack version
0.35.1+git-3a22a7f.build-6099
```

If we set the go.mod to `1.22.0` does it mean the go-setup action will use go 1.22.0?

**Comment by cmoulliard** [comment]
> If we bump the go mod version to `1.22.0` will this affect the workflow version when we are compiling the tool [here](https://github.com/buildpacks/pack/blob/main/.github/workflows/build.yml#L189-L192)?

I'm not a go expert but I don't think so as according to the documentation of the toolchain - https://go.dev/doc/toolchain

"In the standard configuration, the go command uses its own bundled toolchain when that toolchain is at least as new as the go or toolchain lines in the main module or workspace.

For example, when using the go command bundled with Go 1.21.3 in a main module that says go 1.21.0, the go command uses Go 1.21.3. 

When the go or toolchain line is newer than the bundled toolchain, the go command runs the newer toolchain instead. For example, when using the go command bundled with Go 1.21.3 in a main module that says go 1.21.9, the go command finds and runs Go 1.21.9 instead. 

It first looks in the PATH for a program named go1.21.9 and otherwise downloads and caches a copy of the Go 1.21.9 toolchain. 

This automatic toolchain switching can be disabled, but in that case, for more precise forwards compatibility, the go command will refuse to run in a main module or workspace in which the go line requires a newer version of Go. That is, the go line sets the minimum required Go version necessary to use a module or workspace."

**Comment by jjbustamante** [comment]
I will try to test the github action to double check how it behaves

**Comment by cmoulliard** [comment]
> I will try to test the github action to double check how it behaves

Have you been able to test ?

**Comment by jjbustamante** [comment]
@cmoulliard oh man, no! sorry

**Comment by natalieparellano** [comment]
Summarizing other conversation with Juan: we need to check if providing the toolchain in go.mod will lock us to a go version when running actions/setup-go. Ideally we'd like to preserver the ability to automatically upgrade to the latest go patch when we run our pipelines

**Comment by cmoulliard** [comment]
> Ideally we'd like to preserver the ability to automatically upgrade to the latest go patch when we run our pipelines

You can keep the control par of the github action over what you would like to use as go distribution. See: https://github.com/actions/setup-go?tab=readme-ov-file#v3

From a development, testing and releasing point of view, if you decide to align the code to 1.21, then this is a good practice to use now the new convention and to build using latest patch version available if some CVEs, etc have been fixed => example: 1.21.13

**Comment by jjbustamante** [comment]
@cmoulliard I think this issue was fixed on this [PR](https://github.com/buildpacks/pack/pull/2312) `pack 0.36.3` ?

**Comment by cmoulliard** [comment]
> I think this issue was fixed on this [PR](https://github.com/buildpacks/pack/pull/2312) `pack 0.36.3` ?

No. This PR bumps go git versions but not at all the go release version.

Here is an example about what you should do part of the go.mod file: https://github.com/konflux-ci/konflux-ci/blob/main/go.mod#L3-L5
```
go 1.23.0

toolchain go1.23.4
```

**Comment by jjbustamante** [comment]
Sorry, I took the wrong one. 

I mean this [one](https://github.com/buildpacks/pack/pull/2321/files#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6R149)

**Comment by cmoulliard** [comment]
> Sorry, I took the wrong one.

Indeed, this one is better and fixes the issue reported

**Comment by natalieparellano** [comment]
I see that we have `toolchain go1.23.3` now in the go.mod file which is having the effect of locking us to that specific patch version instead of allowing the patch to auto-bump when we run CI. For example the latest build (see [action run](https://github.com/buildpacks/pack/actions/runs/13520701532/job/37786300131#step:6:25)) uses 1.23.3 even though 1.23.6 is available.

Since we're aiming to be explicit we probably want to add some automation that detects a go patch update and PRs the repo to update the toolchain version.
